### PR TITLE
Handle ThrottlingException

### PR DESCRIPTION
**Pull request by uttamk**
##### Problem
When the number of concurrent builds is high, AWS throttles API calls to check for build status and logs. This causes some builds to fail on Jenkins. See https://github.com/awslabs/aws-codebuild-jenkins-plugin/issues/31

##### Solution
Use the AWS JDK provided `RetryPolicy` setting in the ClientConfiguration. The retry policy will have a `RetryCondition` that retries only on throttling exception and not anything else. It will have an `ExponentialBackoffStrategy` that sleeps for increasing periods of time before retrying.

```
This contribution is under the Apache 2.0 license.
```

### CODE REVIEW

File: src/main/java/AWSClientFactory.java [hunk 1]
```diff
@@ -58,12 +62,15 @@ public AWSClientFactory(String proxyHost, String proxyPort, String awsAccessKey,
...
 
-        clientConfig = new ClientConfiguration();
-        clientConfig.setUserAgentPrefix("CodeBuild-Jenkins-Plugin"); //tags all calls made from Jenkins plugin.
         Validation.checkAWSClientFactoryProxyConfig(proxyHost, proxyPort);
-        clientConfig.setProxyHost(proxyHost);
+        clientConfig = new ClientConfiguration()
+                .withUserAgentPrefix("CodeBuild-Jenkins-Plugin") //tags all calls made from Jenkins plugin.
+                .withProxyHost(proxyHost)
+                .withRetryPolicy(new RetryPolicy(new CodeBuildClientRetryCondition(),
+                        new PredefinedBackoffStrategies.ExponentialBackoffStrategy(5000, 20000),
```
Comment (uttamk) [inline src/main/java/AWSClientFactory.java]: These values worked for us when we have around 100 concurrent builds running. Anything less still threw up `ThrottlingException`

**Comment by uttamk** [comment]
@leobaran-aws @oliver-schoenherr let me know what you folks think about this

**Comment by leobaran-aws** [comment]
Thanks for your contribution! The code looks good--I'll test it before merging.

**Comment by leobaran-aws** [comment]
Merged. Thanks again--it's super cool to hear that you're able to run 100 concurrent builds with the plugin!

**Comment by Nevvea7** [comment]
We're planning to run a lot more than 100 concurrent builds, and these tests run for a long time so I don't mind start polling at once a minute or something. @leobaran-aws any chance you can consider adding the pollingInterval option made in #32 to the plugin so users can decide how often to poll?

**Comment by leobaran-aws** [comment]
I'm actually working on that right now by modifying this pull request: https://github.com/jenkinsci/aws-codebuild-plugin/pull/10. I plan to release it in version `0.39` tomorrow. This will allow you to set the minimum polling interval, maximum polling interval, and the polling jitter. For your use case, I'd imagine a minimum poll of 60 seconds and maximum of 90-120 would be sufficient for builds that run for 30+ minutes.

**Comment by Nevvea7** [comment]
Gee that's awesome, thank you so much!

**Comment by leobaran-aws** [comment]
Implemented in 4d45a86990955551a1e53216ec0c98590c088019 which I'll release in version `0.39` today. Let us know how it works for you. You can configure these settings in `Manage Jenkins` > `Configure System` under the `AWS CodeBuild plugin` section (I'll add this to the README as well).

**Comment by Nevvea7** [comment]
Thank you, I tried it on a test Jenkins master and it seems to work, since I need to restart Jenkins to have the update take effect I'll try it out on the other ones tomorrow. 

Just curious, what's the reason for putting this inside `Configure System` and not as params passed in to the step?

**Comment by leobaran-aws** [comment]
Great! The current build-level configuration is pretty complex and I think polling config is globally applicable. But if you or another customer has a use case for build-level polling configuration, we can certainly implement that.

**Pull request closed without merging**